### DISCUSSION TITLE
Pytest uses wrong base directory for tests selection

### DISCUSSION THREAD

**Discussion by SanctusAnimus** [pytest-dev/pytest#10967] [Q&A] [answered]
Suppose i have a following structure, inside Pycharm Project directory
```
ot3_webserver
| - app (folder with code src)
| - ot3_tests
    | - conftest.py
    | - test_routers
        | - test_redis.py

ar_webserver
| - app (folder with code src)
| - tests
    | - conftest.py
    | - test_routers
        | - test_redis.py
```

`ar_webserver` was made initially, and test code was made and written here.
Later tests were migrated (with some adjustments) to `ot3_webserver`

In `ar_webserver` tests still work correctly.
However when i start them in `ot3_webserver` i get following output:
```
test_routers/test_battle_pass_router.py::TestBattlePassRouter::test_redeem_bp_rewards <- ..\tests\test_routers\test_battle_pass_router.py
```

I can't find anything in regards to this `<-` arrow, why it happens and how to prevent that at all.

Inherently, all error messages are pointing towards tests of other projects, like so:
```
..\..\ar_webserver_template\tests\test_routers\test_pages\test_statistics.py:22 - ...
```

Pycharm seems to launch them with correct working directory set, via
```
C:\Users\den_k\PycharmProjects\ot3_webserver\venv\Scripts\python.exe C:/Users/den_k/AppData/Local/JetBrains/Toolbox/apps/PyCharm-P/ch-0/231.8770.66/plugins/python/helpers/pycharm/_jb_pytest_runner.py --path C:\Users\den_k\PycharmProjects\ot3_webserver\ot3_tests -- -s -x -m "not payments" 
Testing started at 19:19 ...
Launching pytest with arguments -s -x -m not payments C:\Users\den_k\PycharmProjects\ot3_webserver\ot3_tests --no-header --no-summary -q in C:\Users\den_k\PycharmProjects\ot3_webserver\ot3_tests

2023-05-06 19:19:34.454 | INFO     | ot3_webserver.ot3_tests.conftest:<module>:13 - CWD: C:\Users\den_k\PycharmProjects\ot3_webserver\ot3_tests
```
But then collected tests all reference external folder
```
test_routers/test_battle_pass_router.py::TestBattlePassRouter::test_set_redeemed <- ..\tests\test_routers\test_battle_pass_router.py 
test_routers/test_battle_pass_router.py::TestBattlePassRouter::test_create_trial_subscription <- ..\tests\test_routers\test_battle_pass_router.py 
test_routers/test_gift_code_router.py::TestGiftCodeRouter::test_prepare_data <- ..\..\ar_webserver_template\tests\test_routers\test_gift_code_router.py 
```

How can i stop pytest from seeking folders above project root?

**Comment by SanctusAnimus** [comment]
As an added detail, it does that only for identical test copies (files that were not changed)
Adding a single comment with whitespace is enough to force it to use test from desired directory

Still have no idea why it even tries to discover tests from folders above working directory though

**Comment by RonnyPfannschmidt** [comment]
This looks like a import path conflict, either add loads of `__init__.py` files to ensure pytest can leave sys. Path alone for imports, or try the importlib import mode

**Reply by SanctusAnimus** [reply]
I believe i have `__init__.py` pretty much everywhere, including every project folder and tests folders themselves.

Can you elaborate to importlib import mode?

**Reply by SanctusAnimus** [reply]
Ok i actually found it and enabled, and it still tries to import wrong files, but now raises when doing so
![image](https://user-images.githubusercontent.com/43399078/236636761-080428a5-085d-4372-bb32-eb3d56d78e7c.png)

If i edit the file to remove said error it'll be gone, but editing file worked without changing import mode anyway

Could it be because of following options of Pycharm messing with Path?
![image](https://user-images.githubusercontent.com/43399078/236636916-18de62ac-abb9-42c4-99c2-ca7d9d9d9586.png)
They are enabled by default, but removing marking from tests folders (they were marked as Test source previously) doesn't seem to do anything, unless i have to reset it somehow
And unchecking them breaks entire test suite cause it fails to find `app` and other modules located on the same level as `tests` folder is

**Reply by RonnyPfannschmidt** [reply]
I'm unfamiliar with the path handling of pycharm, personally I absolutely avoid letting it mess with pythonpath

**Reply by SanctusAnimus** [reply, marked as answer]
Well, for now adding `# ot3_webserver` as a first line to every test file of ot3_webserver project test suite resolved the issue.

Kinda dirty, but as long as it works i guess
